### PR TITLE
Add all agent config to `agent.conf` file for convenient containerization use cases

### PR DESCRIPTION
**Pull request by kezhenxu94**
- [x] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>. NO
- [x] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/CHANGES.md).

### CODE REVIEW

File: CHANGES.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ Release Notes.
 * Support Multiple DNS period resolving mechanism
 * Modify `Tags.STATUS_CODE` field name to `Tags.HTTP_RESPONSE_STATUS_CODE` and type from `StringTag` to `IntegerTag`, add `Tags.RPC_RESPONSE_STATUS_CODE` field to hold rpc response code value.
 * Fix kafka-reporter-plugin shade package conflict
+* Add all config items to `agent.conf` file for convenient containerization use cases.
```
Comment (wu-sheng) [inline CHANGES.md:19]: Let's add another change log about which System ENVs are removed.

Comment (kezhenxu94) [reply CHANGES.md:19]: There is not system env removed, they are sorted to the same orders as in the doc

**Review by wu-sheng — APPROVED** [review]
Once all tests passed, I am good with this.

File: apm-sniffer/config/agent.config [hunk 2]
```diff
@@ -15,100 +15,243 @@
...
+meter.max_meter_size=${SW_METER_MAX_METER_SIZE:500}
+#  Mount the specific folders of the plugins. Plugins in mounted folders would work.
 
-# Match spring bean with regex expression for classname
-# plugin.springannotation.classname_match_regex=${SW_SPRINGANNOTATION_CLASSNAME_MATCH_REGEX:}
+plugin.mount=${SW_MOUNT_FOLDERS:plugins,activations}
+# Peer maximum description limit.
+plugin.peer_max_length=${SW_PLUGIN_PEER_MAX_LENGTH:200}
+# Exclude some plugins define in plugins dir.Plugin names is defined in [Agent plugin list](Plugin-list.md)
+plugin.exclude_plugins=${SW_EXCLUDE_PLUGINS :}
```
Comment (mrproliu) [inline apm-sniffer/config/agent.config:160]: There is an extra space in the configuration item.

File: apm-sniffer/config/agent.config [hunk 3]
```diff
@@ -15,100 +15,243 @@
...
+plugin.kafka.bootstrap_servers=${SW_KAFKA_BOOTSTRAP_SERVERS:localhost:9092}
+#  Timeout period of reading topics from the Kafka server, the unit is second. 
+plugin.kafka.get_topic_timeout=${SW_GET_TOPIC_TIMEOUT:10}
+#  Kafka producer configuration.
+plugin.kafka.consumer_config=${SW_PLUGIN_KAFKA_CONSUMER_CONFIG:}
+# Kafka producer configuration. Read [producer configure](http://kafka.apache.org/24/documentation.html#producerconfigs)
+# to get more details. Check [Kafka report doc](How-to-enable-kafka-reporter.md) for more details and examples.
+plugin.kafka.producer_config=${SW_PLUGIN_KAFKA_PRODUCER_CONFIG:}
+#  Specify which Kafka topic name for Meter System data to report to. 
+plugin.kafka.topic_meter=${SW_PLUGIN_KAFKA_TOPIC_METER:skywalking_meters}
```
Comment (mrproliu) [inline apm-sniffer/config/agent.config:225]: the default value is `skywalking-meters`.

File: apm-sniffer/config/agent.config [hunk 4]
```diff
@@ -15,100 +15,243 @@
...
+#  Specify which Kafka topic name for JVM metrics data to report to. 
+plugin.kafka.topic_metrics=${SW_PLUGIN_KAFKA_TOPIC_METRICS:skywalking_metrics}
```
Comment (mrproliu) [inline apm-sniffer/config/agent.config:227]: the default value is `skywalking-metrics`

File: apm-sniffer/config/agent.config [hunk 5]
```diff
@@ -15,100 +15,243 @@
...
+#  Specify which Kafka topic name for traces data to report to. 
+plugin.kafka.topic_segment=${SW_PLUGIN_KAFKA_TOPIC_SEGMENT:skywalking_segments}
```
Comment (mrproliu) [inline apm-sniffer/config/agent.config:229]: the default value is `skywalking-segments`

File: apm-sniffer/config/agent.config [hunk 6]
```diff
@@ -15,100 +15,243 @@
...
+#  Specify which Kafka topic name for Thread Profiling snapshot to report to. 
+plugin.kafka.topic_profilings=${SW_PLUGIN_KAFKA_TOPIC_PROFILINGS:skywalking_profilings}
```
Comment (mrproliu) [inline apm-sniffer/config/agent.config:231]: the property key should be `plugin.kafka.topic_profiling` and default value is `skywalking-profilings`

File: apm-sniffer/config/agent.config [hunk 7]
```diff
@@ -15,100 +15,243 @@
...
+#  Specify which Kafka topic name for the register or heartbeat data of Service Instance to report to. 
+plugin.kafka.topic_management=${SW_PLUGIN_KAFKA_TOPIC_MANAGEMENT:skywalking_managements}
```
Comment (mrproliu) [inline apm-sniffer/config/agent.config:233]: the default value is `skywalking-managements`

File: apm-sniffer/config/agent.config [hunk 8]
```diff
@@ -15,100 +15,243 @@
...
+#  Apache Dubbo provider collect `arguments` in RPC call, use `Object#toString` to collect `arguments`. 
+plugin.dubbo.collect_provider_arguments=${SW_PLUGIN_DUBBO_COLLECT_PROVIDER_ARGUMENTS:false}
+#  When `plugin.dubbo.provider_consumer_arguments` is `true`, Arguments of length from the front will to the OAP backend 
+plugin.dubbo.consumer_provider_length_threshold=${SW_PLUGIN_DUBBO_CONSUMER_PROVIDER_LENGTH_THRESHOLD:256}
+#  A list of host/port pairs to use for establishing the initial connection to the Kafka cluster. 
+plugin.kafka.bootstrap_servers=${SW_KAFKA_BOOTSTRAP_SERVERS:localhost:9092}
+#  Timeout period of reading topics from the Kafka server, the unit is second. 
+plugin.kafka.get_topic_timeout=${SW_GET_TOPIC_TIMEOUT:10}
+#  Kafka producer configuration.
+plugin.kafka.consumer_config=${SW_PLUGIN_KAFKA_CONSUMER_CONFIG:}
```
Comment (mrproliu) [inline apm-sniffer/config/agent.config:220]: Can't see where is using this config item?

**Review by mrproliu — COMMENTED** [review]
I see some default value is not right with the document, I think we should update too.

File: apm-sniffer/config/agent.config (hunk 8 above)
Comment (wu-sheng) [reply apm-sniffer/config/agent.config:220]: I think it is in `KafkaProducerManager#run`. 

@dmsolr Could you be clear about how to add this property? I mean for a hashmap type.

**Review by wu-sheng — APPROVED** [review]

**Comment by wu-sheng** [comment]
I rerun the Kafka e2e. The Kafka/Meter is always unstable :(

**Comment by wu-sheng** [comment]
@kezhenxu94 Could you check meter e2e? It seems failing again.

**Review by wu-sheng — APPROVED** [review]

**Review by mrproliu — APPROVED** [review]

File: apm-sniffer/config/agent.config (hunk 8 above)
Comment (dmsolr) [reply apm-sniffer/config/agent.config:220]: It is my mistake. The proper configuration item is `plugin.kafka.producer_config`.

The format is
```
plugin.kafka.producer_config[key]=value
```
and the key list was [here](https://kafka.apache.org/23/documentation.html#producerconfigs). 
I do not try to set it in JVM Options. It seems that you can't do it.

Comment (wu-sheng) [reply apm-sniffer/config/agent.config:220]: @kezhenxu94 Any preference to do this? This mode is hard to override through system env. We can't list them all, they are too many.

Comment (kezhenxu94) [reply apm-sniffer/config/agent.config:220]: > @kezhenxu94 Any preference to do this? This mode is hard to override through system env. We can't list them all, they are too many.

What about avoiding the key in the config key, for example `plugin.kafka.producer_config=key1=value1,key2=value2`, if we could use JSON in agent, we may use `plugin.kafka.producer_config={"key":"val"}`?

Comment (wu-sheng) [reply apm-sniffer/config/agent.config:220]: > plugin.kafka.producer_config=key1=value1,key2=value2

I prefer to support this first in the agent initialization through another PR? @dmsolr Could you do this?

Comment (dmsolr) [reply apm-sniffer/config/agent.config:220]: > plugin.kafka.producer_config=key1=value1,key2=value2
It is good to me.

> @dmsolr Could you do this?
Sure.

Comment (wu-sheng) [reply apm-sniffer/config/agent.config:220]: Please create an issue to track this.

**Comment by dashanji** [comment]
I find some configurations are in configurations.md , but not in the agent.config , such as ```agent.trace_segment_ref_limit_per_span``` 、```agent.instance_properties[key]=value``` . Otherwise , the configuration ```plugin.dubbo.consumer_provider_length_threshold``` in the agent.config is not the same as ```plugin.dubbo.provider_arguments_length_threshold``` in the configurations.md.

**Comment by wu-sheng** [comment]
> I find some configurations are in configurations.md , but not in the agent.config , such as `agent.trace_segment_ref_limit_per_span` 、`agent.instance_properties[key]=value` . Otherwise , the configuration `plugin.dubbo.consumer_provider_length_threshold` in the agent.config is not the same as `plugin.dubbo.provider_arguments_length_threshold` in the configurations.md.

If there are names mismatch or missing, submit a pull request to fix directly. 
Things like `agent.instance_properties[key]=value` are not pre-defined by SkyWalking, we don't know what to add.

**Comment by kezhenxu94** [comment]
> I find some configurations are in configurations.md , but not in the agent.config , such as ```agent.trace_segment_ref_limit_per_span``` 、```agent.instance_properties[key]=value``` . Otherwise , the configuration ```plugin.dubbo.consumer_provider_length_threshold ``` in the agent.config is not the same as ```plugin.dubbo.provider_arguments_length_threshold``` in the configurations.md.  

@dashanji can you open a pull request to fix this? For `agent.instance_properties[xxx]`, I think we also need to migrate to the format of `agent.instance_properties_json`.  `plugin.dubbo.consumer_provider_length_threshold` should be typo of `plugin.dubbo.consumer_length_threshold`

**Comment by dashanji** [comment]
> > I find some configurations are in configurations.md , but not in the agent.config , such as `agent.trace_segment_ref_limit_per_span` 、`agent.instance_properties[key]=value` . Otherwise , the configuration `plugin.dubbo.consumer_provider_length_threshold ` in the agent.config is not the same as `plugin.dubbo.provider_arguments_length_threshold` in the configurations.md.
> 
> @dashanji can you open a pull request to fix this? For `agent.instance_properties[xxx]`, I think we also need to migrate to the format of `agent.instance_properties_json={}`. `plugin.dubbo.consumer_provider_length_threshold` should be typo of `plugin.dubbo.consumer_length_threshold`

OK.

**Comment by dashanji** [comment]
> > I find some configurations are in configurations.md , but not in the agent.config , such as `agent.trace_segment_ref_limit_per_span` 、`agent.instance_properties[key]=value` . Otherwise , the configuration `plugin.dubbo.consumer_provider_length_threshold` in the agent.config is not the same as `plugin.dubbo.provider_arguments_length_threshold` in the configurations.md.
> 
> If there are names mismatch or missing, submit a pull request to fix directly.
> Things like `agent.instance_properties[key]=value` are not pre-defined by SkyWalking, we don't know what to add.

OK , I will open a pull request to fix it.

**Comment by wu-sheng** [comment]
`agent.instance_properties_json={}` should be a new feature, rather than a fix. @dashanji If you like, please submit another new(not the wrong names fix) PR to add this.

**Comment by dashanji** [comment]
> `agent.instance_properties_json={}` should be a new feature, rather than a fix. @dashanji If you like, please submit another new(not the wrong names fix) PR to add this.

Sorry I don't know much about java, so I can't realize this feature at present.

**Comment by wu-sheng** [comment]
> > `agent.instance_properties_json={}` should be a new feature, rather than a fix. @dashanji If you like, please submit another new(not the wrong names fix) PR to add this.
> 
> Sorry I don't know much about java, so I can't realize this feature at present.

OK, then if we need someone to add this feature first, otherwise, `agent.instance_properties[key]=value` can't be added, just purely because we don't know what to add.

But others you proposed, feel free to send a pull request to fix.

**Pull request merged**